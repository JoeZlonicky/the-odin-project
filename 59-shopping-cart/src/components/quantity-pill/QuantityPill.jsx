import Button from '../button/Button';
import styles from './QuantityPill.module.css';

function QuantityPill({ quantity, setQuantity, id }) {
  const decrement = () => {
    if (quantity <= 0) return;
    setQuantity(quantity - 1);
  };

  const increment = () => {
    if (quantity >= 999) return;
    setQuantity(quantity + 1);
  };

  const onInputChange = (newValue) => {
    newValue = newValue.replace(/\D/g, '');
    if (newValue == '') {
      setQuantity(0);
      return;
    }

    newValue = parseInt(newValue);
    setQuantity(newValue);
  };

  return (
    <div className={styles.quantityPill}>
      <label className={styles.label} htmlFor={id + '-input'}>
        Quantity
      </label>
      <div className={styles.inputContainer}>
        <Button onClick={decrement} ariaLabel="Decrement">
          â€”
        </Button>
        <input
          id={id + '-input'}
          min="0"
          className={styles.input}
          value={quantity}
          onChange={(e) => onInputChange(e.target.value)}
          maxLength={3}
        ></input>
        <Button onClick={increment} ariaLabel="Increment">
          +
        </Button>
      </div>
    </div>
  );
}

export default QuantityPill;
