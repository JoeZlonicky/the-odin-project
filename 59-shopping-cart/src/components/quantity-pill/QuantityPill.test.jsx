import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';
import QuantityPill from './QuantityPill';

async function getFunctionCallFromButtonClick(buttonLabelText, initial) {
  const user = userEvent.setup();
  const setQuantity = vi.fn();

  render(<QuantityPill quantity={initial} setQuantity={setQuantity} />);
  const button = screen.getByLabelText(buttonLabelText);
  await user.click(button);

  return setQuantity;
}

async function getFunctionCallFromInput(inputValue, initial) {
  const user = userEvent.setup();
  const setQuantity = vi.fn();

  render(<QuantityPill quantity={initial} setQuantity={setQuantity} />);
  const button = screen.getByRole('textbox');
  await user.type(button, inputValue);

  return setQuantity;
}

it('Matches screenshot', () => {
  const component = render(<QuantityPill />);

  expect(component).toMatchSnapshot();
});

it('Uses quantity', () => {
  const quantity = 3;

  render(<QuantityPill quantity={quantity} />);

  expect(screen.getByRole('textbox').value).toMatch(`${quantity}`);
});

it('Decrements when pressed', async () => {
  const fn = await getFunctionCallFromButtonClick('Decrement', 3);
  expect(fn).toHaveBeenCalledWith(2);
});

it('Increments when pressed', async () => {
  const fn = await getFunctionCallFromButtonClick('Increment', 3);
  expect(fn).toHaveBeenCalledWith(4);
});

it('Doesn\t decrement below 0', async () => {
  const fn = await getFunctionCallFromButtonClick('Decrement', 0);
  expect(fn).not.toHaveBeenCalledWith(-1);
});

it('Doesn\t go above 3 numbers', async () => {
  const fn = await getFunctionCallFromButtonClick('Increment', 999);
  expect(fn).not.toHaveBeenCalled(1000);
});

it('Doesn\t allow alpha input', async () => {
  const fn = await getFunctionCallFromInput('Hello9', 0);
  expect(fn).toHaveBeenCalledWith(9);
});
