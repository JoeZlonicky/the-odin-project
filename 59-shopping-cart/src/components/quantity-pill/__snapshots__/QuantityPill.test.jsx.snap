// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QuantityPill > Matches screenshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_quantityPill_54d314"
      >
        <label
          class="_label_54d314"
          for="undefined-input"
        >
          Quantity
        </label>
        <div
          class="_inputContainer_54d314"
        >
          <button
            aria-label="Decrement"
            class="_button_4fb095"
            type="button"
          >
            —
          </button>
          <input
            class="_input_54d314"
            id="undefined-input"
            maxlength="3"
            min="0"
            value=""
          />
          <button
            aria-label="Increment"
            class="_button_4fb095"
            type="button"
          >
            +
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_quantityPill_54d314"
    >
      <label
        class="_label_54d314"
        for="undefined-input"
      >
        Quantity
      </label>
      <div
        class="_inputContainer_54d314"
      >
        <button
          aria-label="Decrement"
          class="_button_4fb095"
          type="button"
        >
          —
        </button>
        <input
          class="_input_54d314"
          id="undefined-input"
          maxlength="3"
          min="0"
          value=""
        />
        <button
          aria-label="Increment"
          class="_button_4fb095"
          type="button"
        >
          +
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
